homeassistant/auth/__init__.py

```diff
@@ -479,6 +492,31 @@ async def async_remove_refresh_token(
         for revoke_callback in callbacks:
             revoke_callback()
 
+    async def async_track_next_refresh_token_expiration(self) -> None:
+        """Initialise all token expiration scheduled tasks."""
+        next_expiration = dt_util.utcnow() + REFRESH_TOKEN_EXPIRATION
+        for token in await self._store.async_get_refresh_tokens():
+            if (
+                expire_at := token.expire_at
+            ) is not None and expire_at < next_expiration:
+                next_expiration = expire_at
+
+        async def _remove_expired_refresh_tokens(_: datetime | None = None) -> None:
+            now = dt_util.utcnow()
+            for token in list(await self._store.async_get_refresh_tokens()):
```

Comment:
```suggestion
            for token in (await self._store.async_get_refresh_tokens())[:]:
```

Reply:
just to get it - what is the advantage or difference of your suggestion?

Reply:
List makes a new list and the other is a view of the current list

Reply:
ok, thanks for explanation, I've updated the PR

Reply:
> List makes a new list and the other is a view of the current list
> 
> The view is a bit faster

I don't believe a slice of all the items is just a view of the current list.

```
Python 3.11.7 (main, Dec  8 2023, 18:56:57) [GCC 9.4.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: test = [1, 2, 3]

In [2]: test_2 = test[:]

In [3]: test is test_2
Out[3]: False

In [4]: test_2
Out[4]: [1, 2, 3]

In [5]: test
Out[5]: [1, 2, 3]

In [6]: type(test_2)
Out[6]: list

In [7]: test_2.append(4)

In [8]: test_2
Out[8]: [1, 2, 3, 4]

In [9]: test
Out[9]: [1, 2, 3]

```

Reply:
> List makes a new list and the other is a view of the current list
> 
> The view is a bit faster

I don't believe a full slice of a list is just a view of the current list.

```
Python 3.11.7 (main, Dec  8 2023, 18:56:57) [GCC 9.4.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: test = [1, 2, 3]

In [2]: test_2 = test[:]

In [3]: test is test_2
Out[3]: False

In [4]: test_2
Out[4]: [1, 2, 3]

In [5]: test
Out[5]: [1, 2, 3]

In [6]: type(test_2)
Out[6]: list

In [7]: test_2.append(4)

In [8]: test_2
Out[8]: [1, 2, 3, 4]

In [9]: test
Out[9]: [1, 2, 3]
```

Reply:
```python
list1 = ["a", "b", "c", "d"]
list2 = ["a", "b", "c", "d"]
for item in list1:
    list1.remove(item)
for item in list2[:]:
    list2.remove(item)

print(list1)
print(list2)
```
```
% python3 test.py
['b', 'd']
[]
```

Yes python calls it a slice (I think of it a view).  It behaves the way we want with when iterating.

Regardless, spot on that `async_get_refresh_tokens` makes a new list anyways we don't need it anyways.

Reply:
The full slice will make a shallow copy, which is the same as list.copy does.

```
In [1]: list1 = ["a", "b", "c", "d"]
   ...: list2 = ["a", "b", "c", "d"]
   ...: for item in list1:
   ...:     list1.remove(item)
   ...: for item in list2.copy():
   ...:     list2.remove(item)
   ...: 
   ...: 
   ...: print(list1)
   ...: print(list2)
['b', 'd']
[]

In [11]: list1 = ["a", "b", "c", "d"]
    ...: list2 = ["a", "b", "c", "d"]
    ...: list3 = list2[:]
    ...: for item in list1:
    ...:     list1.remove(item)
    ...: for item in list3:
    ...:     list2.remove(item)
    ...: print(list1)
    ...: print(list2)
    ...: print(list3)
['b', 'd']
[]
['a', 'b', 'c', 'd']

Reply:
Looks like `copy()` is better on newer python (but after faster now on 3.12)


py3.12
>>> timeit.timeit('x.copy()',setup='x=["item"*500]')
0.033824500162154436
>>> timeit.timeit('x[:]',setup='x=["item"*500]')
0.036731207743287086
>>> 

py3.6
>>> timeit.timeit('x.copy()',setup='x=["item"*500]')
0.046285499999996205
>>> timeit.timeit('x[:]',setup='x=["item"*500]')
0.043951999999999884

Reply:
actually thats not a good test since I didn't [] in the setup

Reply:
3.12
```
>>> import timeit
>>> timeit.timeit('x[:]',setup='x=["item"]*500')
0.571478916797787
>>> timeit.timeit('x.copy()',setup='x=["item"]*500')
0.5659893746487796
>>> 
```


3.8
```
>>> timeit.timeit('x[:]',setup='x=["item"]*500')
1.0010294579999997
>>> timeit.timeit('x.copy()',setup='x=["item"]*500')
0.984839333
```

Looks like copy is always faster on modern python

Reply:
Also pretty surprising how much better python 3.12 performs

3.11:
```
>>> import timeit
>>> timeit.timeit('x[:]',setup='x=["item"]*500')
1.0051707499660552
>>> timeit.timeit('x.copy()',setup='x=["item"]*500')
0.9945869580842555
```